### ISSUE TITLE
[META] Baseline MAINTAINERS, CODEOWNERS and external collaborator permissions

### ISSUE THREAD

**Issue by dblock** [opensearch-project/.github#125]
### What are you proposing?

There are some common differences across repos in opensearch-project when it comes to maintainers, codeowners and repository permissions.

1. The MAINTAINERS.md list is not the folks actively engaged in a project.
1. PRs get approved by folks that are not in the MAINTAINERS list, because permissions aren't in sync. 
1. CODEOWNERS doesn't exist, so maintainers don't always get properly notified about pull requests or their list doesn't appear in reviewers when contributors make pull requests.
1. CODEOWNERS is out of sync, and does not include a subset, or all maintainers.
1. Actual repository permissions may not include folks in MAINTAINERS as external collaborators.
1. Repo permissions may give some groups access to more repos than they are a member of.

The proposal is to correct and align on all the issues above in a single campaign.

### What problems are you trying to solve?

- Collaborators in opensearch-project expect transparency in who maintainers on any of the projects are. 
- Maintainers want to be notified (or have the option not to be notified) when PRs are opened.
- Maintainers need reliable lists to find other maintainers to be able to follow our process of adding new maintainers.
- Maintainers want to be publicly recognized for their work.

### What will it take to execute?

First, baseline MAINTAINERS if you've never done that.

1. Email/reach out to current maintainers and ask them whether they want to still be a maintainer.
4. Check the list of people in MAINTAINERS.md to include only active maintainers.
6. Add/move non-active maintainers to an emeritus section.
7. Add active maintainers to CODEOWNERS.
8. Remove any broad teams from CODEOWNERS.
9. Make a PR with the changes above, make sure CODEOWNERS is green.
10. Once the PR is approved, ask an org admin or an admin delegate (e.g. @bbarani) to sync the repo permissions, removing groups and adding individual contributors. Let's promote some new folks to this role to help, too.
11. Merge the PR.
12. Maintain the MAINTAINERS.md and CODEOWNERS.md accurate.

### Verified

- [x] https://github.com/opensearch-project/alerting/issues/793
- [x] https://github.com/opensearch-project/anomaly-detection/issues/808
- [x] https://github.com/opensearch-project/asynchronous-search/issues/235
- [x] https://github.com/opensearch-project/common-utils/issues/357
- [x] https://github.com/opensearch-project/index-management-dashboards-plugin/issues/610
- [x] https://github.com/opensearch-project/opensearch-cli/issues/70
- [x] https://github.com/opensearch-project/perftop/issues/64
- [x] https://github.com/opensearch-project/piped-processing-language/issues/18
- [x] https://github.com/opensearch-project/project-website/issues/1354
- [x] https://github.com/opensearch-project/security-analytics-dashboards-plugin/issues/409
- [x] https://github.com/opensearch-project/neural-search/issues/111
- [x] https://github.com/opensearch-project/dashboards-notifications/issues/31
- [x] https://github.com/opensearch-project/alerting-dashboards-plugin/issues/481
- [x] https://github.com/opensearch-project/notifications/issues/616
- [x] https://github.com/opensearch-project/geospatial/issues/220
- [x] https://github.com/opensearch-project/opensearch-migrations/issues/106
- [x] https://github.com/opensearch-project/dashboards-query-workbench/issues/41
- [x] https://github.com/opensearch-project/project-website-search/issues/44
- [x] https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/issues/421
- [x] https://github.com/opensearch-project/index-management/issues/684
- [x] https://github.com/opensearch-project/k-NN/issues/761
- [x] https://github.com/opensearch-project/cross-cluster-replication/issues/704
- [x] https://github.com/opensearch-project/security-analytics/issues/324
- [x] https://github.com/opensearch-project/opensearch-testcontainers/issues/67
- [x] https://github.com/opensearch-project/reporting/issues/643
- [x] https://github.com/opensearch-project/dashboards-visualizations/issues/163
- [x] https://github.com/opensearch-project/performance-analyzer-rca/issues/293
- [x] https://github.com/opensearch-project/sql/issues/1337
- [x] https://github.com/opensearch-project/observability/issues/1407
- [x] https://github.com/opensearch-project/opensearch-plugin-template-java/issues/49
- [x] https://github.com/opensearch-project/maps/issues/37
- [x] https://github.com/opensearch-project/opensearch-catalog/issues/9
- [x] https://github.com/opensearch-project/simple-schema/issues/69
- [x] https://github.com/opensearch-project/sql-cli/issues/10
- [x] https://github.com/opensearch-project/sql-jdbc/issues/57
- [x] https://github.com/opensearch-project/sql-odbc/issues/31
- [x] https://github.com/opensearch-project/dashboards-observability/issues/275
- [x] https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3426
- [x] https://github.com/opensearch-project/i18n-plugin/issues/17
- [x] https://github.com/opensearch-project/data-prepper/issues/2275
- [x] https://github.com/opensearch-project/documentation-website/issues/2878
- [x] https://github.com/opensearch-project/oui/issues/305
- [x] https://github.com/opensearch-project/opensearch-py-ml/issues/81
- [x] https://github.com/opensearch-project/oui-docs-cdk/issues/8
- [x] https://github.com/opensearch-project/performance-analyzer/issues/382
- [x] https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/526
- [x] https://github.com/opensearch-project/opensearch-dashboards-test-library/issues/29
- [x] https://github.com/opensearch-project/opensearch-oci-object-storage/issues/28
- [x] https://github.com/opensearch-project/dashboards-search-relevance/issues/146
- [x] https://github.com/opensearch-project/opensearch-clients/issues/51
- [x] https://github.com/opensearch-project/opensearch-rs/issues/123
- [x] https://github.com/opensearch-project/opensearch-net/issues/151
- [x] https://github.com/opensearch-project/opensearch-api-specification/issues/77
- [x] https://github.com/opensearch-project/opensearch-hadoop/issues/94
- [x] https://github.com/opensearch-project/job-scheduler/issues/316
- [x] https://github.com/opensearch-project/opensearch-devops/issues/111
- [x] https://github.com/opensearch-project/opensearch-py/issues/289
- [x] https://github.com/opensearch-project/ml-commons/issues/727
- [x] https://github.com/opensearch-project/opensearch-dsl-py/issues/96
- [x] https://github.com/opensearch-project/dashboards-docker-images/issues/10
- [x] https://github.com/opensearch-project/opensearch-build-libraries/issues/140
- [x] https://github.com/opensearch-project/OpenSearch/issues/6319
- [x] https://github.com/opensearch-project/opensearch-build/issues/3200
- [x] https://github.com/opensearch-project/opensearch-plugins/issues/205
- [x] https://github.com/opensearch-project/security/issues/2437
- [x] https://github.com/opensearch-project/security-dashboards-plugin/issues/1326
- [x] https://github.com/opensearch-project/project-meta/issues/83
- [x] https://github.com/opensearch-project/opensearch-java/issues/363
- [x] https://github.com/opensearch-project/logstash-output-opensearch/issues/196
- [x] https://github.com/opensearch-project/helm-charts/issues/381
- [x] https://github.com/opensearch-project/ansible-playbook/issues/114
- [x] https://github.com/opensearch-project/opensearch-ci/issues/249
- [x] https://github.com/opensearch-project/dashboards-anywhere/issues/151
- [x] https://github.com/opensearch-project/opensearch-sdk-java/issues/448
- [x] https://github.com/opensearch-project/opensearch-go/issues/230
- [x] https://github.com/opensearch-project/opensearch-js/issues/385
- [x] https://github.com/opensearch-project/opensearch-php/issues/123
- [x] https://github.com/opensearch-project/opensearch-ruby/issues/142
- [x] https://github.com/opensearch-project/opensearch-benchmark/issues/217
- [x] https://github.com/opensearch-project/search-processor/issues/101
- [x] https://github.com/opensearch-project/opensearch-cluster-cdk/issues/12
- [x] https://github.com/opensearch-project/project-tools/issues/51
- [x] https://github.com/opensearch-project/terraform-provider-opensearch/issues/23
- [x] https://github.com/opensearch-project/logstash-input-opensearch/issues/45
- [x] https://github.com/opensearch-project/dashboards-desktop/issues/35
- [x] https://github.com/opensearch-project/opensearch-benchmark-workloads/issues/61
- [x] https://github.com/opensearch-project/spring-data-opensearch/issues/95

**Comment by wbeckler** [comment]
This looks great to me. Reminder to maintainers: review https://github.com/opensearch-project/.github/blob/main/RESPONSIBILITIES.md#becoming-a-maintainer in order to execute these activities according to the agreed policies.

**Comment by CEHENKLE** [comment]
So if the problem is transparency, adding people as "collaborators" does not solve the problem.  

AFAIK, Only admins can see the "setting" tabs where collaborators live, so people will have to trust that codeowners/maintainers is correct (as opposed to today where org member can at least inspect the teams).  Put another way, you haven't added a mechanism to codeowners/maintainers up-to-date, and you've reduced the number of people who can audit it.

The benefit of teams is that we can make someone a "team maintainer" (with the ability to add and remove team members) without giving them any other additional permissions on the repo.  The unintended consequence for moving to collaborators  is that you have have to be an admin to add folks.  Because of all the other things admins can do, I want to keep the number of admins down to a very small number.   

Now if the problem we're solving is "maintainers are getting added, and we want to slow that process down and add a gate to it" (which is a reasonable problem), then making it an admin-only function will do that. 

Now an interesting thing would be to say the ONLY way you can get added as a collaborator is if you're in codeowners/maintainers MD (e.g. tooling that automatically syncs one to the other).  There's tooling out there to remove folks, so presumably we could have tooling to add folks as well...?

**Comment by dblock** [comment]
@CEHENKLE Don't you have to be a member of the org to be added to a team? Assuming that's possible, isn't that the same problem of admin gating?

**Comment by dlvenable** [comment]
@dblock , What is the motivation for using a list of maintainers in the CODEOWNERS file rather than the team identifier? The current approach has one fewer place to have to worry about when updating the current maintainers list.

With this change, there are three things to change: 1) MAINTAINERS.md, 2) CODEOWNERS, 3) Team membership

**Comment by dblock** [comment]
> @dblock , What is the motivation for using a list of maintainers in the CODEOWNERS file rather than the team identifier? The current approach has one fewer place to have to worry about when updating the current maintainers list.

Teams are useful when granting r/w access to several repos.

- You cannot have external contributors in a team without adding them to the org, which is not currently possible.
- Teams are being added left and right to repos, then people are added to teams granting them rights to repos without updating MAINTAINERS. 
- You cannot see team members from the outside.

So I propose we get rid of teams to treat org members the same way as external collaborators.

The nice thing about teams is that you can have admins of a team without being an admin of a repo. But the above arguments make it impractical.

> With this change, there are three things to change: 1) MAINTAINERS.md, 2) CODEOWNERS, 3) Team membership

I propose to get rid of teams altogether.

**Comment by dlvenable** [comment]
Thanks for the detailed explanation! The current problems make sense and I understand why repo permissions can help resolve this.

> > With this change, there are three things to change: 1) MAINTAINERS.md, 2) CODEOWNERS, 3) Team membership
>
> I propose to get rid of teams altogether.

We will still have three places to change. Instead of the Team membership, we need to update the repo permissions. But, I understand the problem now so I'm onboard with this.

**Comment by gaiksaya** [comment]
One more use case is here https://github.com/opensearch-project/opensearch-build/issues/3111

**Comment by wbeckler** [comment]
Given that the rules on maintainers are repo-wide and consistent, could we get rid of codeowners and modify settings so that it is no longer needed?

**Comment by dblock** [comment]
> Given that the rules on maintainers are repo-wide and consistent, could we get rid of codeowners and modify settings so that it is no longer needed?

What are you proposing specificially?

**Comment by dbwiddis** [comment]
> 7. Once the PR is approved, ask an org admin or an admin delegate (e.g. @dblock) to help you sync the repo permissions, removing groups and adding individual contributors. Let's promote some new folks to this role to help, too.

I've removed all groups except the project-wide "admin" group.  Is that intended to be left, or is it included in "removing groups" as the repo maintainers won't have visibility to update the permissions if that (org level) group changes.

**Comment by wbeckler** [comment]
> > Given that the rules on maintainers are repo-wide and consistent, could we get rid of codeowners and modify settings so that it is no longer needed?
> 
> What are you proposing specifically?

I'm proposing that either: 

1. We have a github action that copies the contents of CODEOWNERS into Maintainers.md or vice versa, or
2. We get rid of CODEOWNERS and change the branch protection settings to rely on the maintainers list (if possible), or
3. We get rid of Maintainers.md and point the CONTRIBUTING.md to the CODEOWNERS as the list of maintainers.

The idea is to DRY these two files.

**Comment by dbwiddis** [comment]
> I'm proposing that either:
> 
> 1. We have a github action that copies the contents of CODEOWNERS into Maintainers.md or vice versa, or

CODEOWNERS has a very specific format that doesn't include fields we have in maintainers such as affiliation.  Copying the other way requires a lot of parsing.  I'm not sure either file changes often enough that the [juice is worth the squeeze](https://xkcd.com/1205/) on automation.

> 2. We get rid of CODEOWNERS and change the branch protection settings to rely on the maintainers list (if possible), or

CODEOWNERS is tightly integrated into GitHub and is directly used in branch protection rules.  For example, our repo requires a PR be approved by a code owner.  It's not really practical/possible to include such permissions with another file.

> 3. We get rid of Maintainers.md and point the CONTRIBUTING.md to the CODEOWNERS as the list of maintainers.

Reasonable if we don't want to include "Affiliation", but also drops the notion of "emeritus" maintainers or ability to acknowledge contributors who are no longer active but are welcome to return from a hiatus.

**Comment by wbeckler** [comment]
I agree with @dbwiddis and have no further ideas.

**Comment by dblock** [comment]
@wbeckler Do you think we can close this or do you still think we should do something?

**Comment by anasalkouz** [comment]
> Check the list of people in MAINTAINERS.md to include only active maintainers.
Add/move non-active maintainers to an emeritus section.

What is the criteria to define someone as an active or not

> Add active maintainers to CODEOWNERS.

Not sure what could be the case where you have someone in MAINTAINERS.md but not in the CODEOWNERS. if there is no significant difference and mostly they will be an identical list, then we why we need to maintain 2 lists. How to make sure CODEOWNERS include a subset or all maintainers.

**Comment by dbwiddis** [comment]
> What is the criteria to define someone as an active or not

In our case we moved someone to Emeritus in MAINTAINTERS but removed them from CODEOWNERS.

> then we why we need to maintain 2 lists

CODEOWNERS is tightly integrated into GitHub PR review requests and branch protection features and contains only usernames (and optionally split them among file types or subdirectories; in most cases we all do `*`).  We can't really get rid of this one.

MAINTAINERS is freeform and i allows additional information such as "Emeritus" and employer affiliation.  In theory we could also add more information like specializations within the project for people reaching out.  If you want to reduce redundancy you'd eliminate this one.

**Comment by macohen** [comment]
Is there a way to contact all admins (e.g. create a new issue) for "What will it take to execute?" step #7. Otherwise, I think everyone will just ask @dblock and he'll end up doing all the work there.

**Comment by joshuali925** [comment]
I'm working on all new issues for sql repos and just noticed this one. Item 3 ([ref](https://github.com/opensearch-project/reporting/issues/643)) says

> Repo permissions only contain individual aliases as collaborators with maintain rights, admin, and triage teams.

I don't have admin access, would I still be able to do this? This page https://github.com/opensearch-project/reporting/settings/access is 404 for me. If I can't check this, who should be doing it?

**Comment by dblock** [comment]
> > Check the list of people in MAINTAINERS.md to include only active maintainers.
> > Add/move non-active maintainers to an emeritus section.
> 
> What is the criteria to define someone as an active or not
> 
> > Add active maintainers to CODEOWNERS.
> 
> Not sure what could be the case where you have someone in MAINTAINERS.md but not in the CODEOWNERS. if there is no significant difference and mostly they will be an identical list, then we why we need to maintain 2 lists. How to make sure CODEOWNERS include a subset or all maintainers. @dblock

When you merge a change to CODEOWNERS, GitHub will tell you whether it's valid (a subset of people with actual permissions).

**Comment by dblock** [comment]
> > Check the list of people in MAINTAINERS.md to include only active maintainers.
> > Add/move non-active maintainers to an emeritus section.
> 
> What is the criteria to define someone as an active or not
> 
> > Add active maintainers to CODEOWNERS.
> 
> Not sure what could be the case where you have someone in MAINTAINERS.md but not in the CODEOWNERS. if there is no significant difference and mostly they will be an identical list, then we why we need to maintain 2 lists. How to make sure CODEOWNERS include a subset or all maintainers. @dblock

When you merge a change to CODEOWNERS, GitHub will tell you whether it's valid (a subset of people with actual permissions).

**Comment by dblock** [comment]
> Is there a way to contact all admins (e.g. create a new issue) for "What will it take to execute?" step #7. Otherwise, I think everyone will just ask @dblock and he'll end up doing all the work there.

I think tagging @opensearch-project/admin on a PR and asking them to confirm permission changes is the easiest way.

**Comment by bbarani** [comment]
Created missing issue for docker-images - https://github.com/opensearch-project/docker-images/issues/21

**Comment by dblock** [comment]
> Would it be helpful to have a central tool that automatically sends baseline emails to get confirmation on who wants to be a maintainer or not for all the repos?

You mean some kind of automation that asks a person whether they want to be a maintainer? Personally I didn't need something like this. In the past month I've nominated 2 maintainers, so had to send a total of 4 emails (2 to existing maintainers and then one to the person I nominated) and open 2 PRs. It's really not that much.

**Comment by lezzago** [comment]
> You mean some kind of automation that asks a person whether they want to be a maintainer? Personally I didn't need something like this. In the past month I've nominated 2 maintainers, so had to send a total of 4 emails (2 to existing maintainers and then one to the person I nominated) and open 2 PRs. It's really not that much.

This is for baselining current maintainers. This way we can automate the baselining procedure to ask all maintainers every so often (the decided interval) if they want to continue being a maintainer. This is to ensure all repos are following the correct baselining procedure.

**Comment by dblock** [comment]
@lezzago We introduced emeritus to allow maintainers to say that they no longer want to be involved, and to still recognize their past contributions. How would actively removing folks from MAINTAINERS because they went inactive have any material impact? 

FYI I am planning to extend https://github.com/opensearch-project/project-tools `maintainers audit` to verify that permissions match what we have in CODEOWNERS and MAINTAINERS. I definitely think it would be useful to add something like "how long ago did a maintainer contribute last?".

**Comment by dbwiddis** [comment]
>  "how long ago did a maintainer contribute last?".

What is the definition of "contribute"?

**Comment by seanneumann** [comment]
**(thanks to @ananzh)**
[documentation-website#2878](https://github.com/opensearch-project/documentation-website/issues/2878) (PR: https://github.com/opensearch-project/documentation-website/pull/3639 and https://github.com/opensearch-project/documentation-website/pull/3641, Status: In review)
O[penSearch-Dashboards#3426](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3426) (PR: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3792, Status: In review)
[opensearch-dashboards-functional-test#526](https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/526) (PR: https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/606, Status: In review)
[oui#305](https://github.com/opensearch-project/oui/issues/305) (PR: https://github.com/opensearch-project/oui/pull/668, Status: In review)
[i18n-plugin#17](https://github.com/opensearch-project/i18n-plugin/issues/17) (PR: https://github.com/opensearch-project/i18n-plugin/pull/20, Status: merged)
[opensearch-catalog#9](https://github.com/opensearch-project/opensearch-catalog/issues/9) (PR: https://github.com/opensearch-project/opensearch-catalog/pull/11 Status: PR is out but I don’t see a merge button)
[oui-docs-cdk#8](https://github.com/opensearch-project/oui-docs-cdk/issues/8) (PR: https://github.com/opensearch-project/oui-docs-cdk/pull/10 Status: PR is out but I don’t see a merge button)
Done and update each one with PR and status. I don’t see a merge button for the last two. Maybe I don’t have a permission.

**Comment by peternied** [comment]
@seanneumann Here is a cleaner version of your list

- https://github.com/opensearch-project/documentation-website/issues/2878
  - https://github.com/opensearch-project/documentation-website/pull/3639
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3426
  - https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3792
- https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/526
  - https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/606
- https://github.com/opensearch-project/oui/issues/305
  - https://github.com/opensearch-project/oui/pull/668
- https://github.com/opensearch-project/i18n-plugin/issues/17
  - https://github.com/opensearch-project/i18n-plugin/pull/20
- https://github.com/opensearch-project/opensearch-catalog/issues/9
  - https://github.com/opensearch-project/opensearch-catalog/pull/11
- https://github.com/opensearch-project/oui-docs-cdk/issues/8
  - https://github.com/opensearch-project/oui-docs-cdk/pull/10


Trick is to just reference the url of the issue/PR in list form, e.g.
```
- https://github.com/opensearch-project/oui-docs-cdk/issues/8
  - https://github.com/opensearch-project/oui-docs-cdk/pull/10
```

**Comment by seanneumann** [comment]
Thanks @peternied!

**Comment by dbwiddis** [comment]
> Thanks @peternied! I'm a little to quick to copy and paste. :)

But do you have [_The Key_](https://stackoverflow.blog/2021/03/31/the-key-copy-paste/) on your desk like I do?

**Comment by SuZhou-Joe** [comment]
There is no such Baseline issue in https://github.com/opensearch-project/dashboards-notifications/issues, @dblock could you please sync up one on this repo as well?

**Comment by macohen** [comment]
In this scheme, should maintainers should be able to add new maintainers to the repositories without intervention from the opensearch-project/admins? I thought yes, but I just reopened https://github.com/opensearch-project/dashboards-search-relevance/issues/146 because I'm unable to access Settings -> Collaborators on the repo.

**Comment by dblock** [comment]
> There is no such Baseline issue in [dashboard-notifications](https://github.com/opensearch-project/dashboards-notifications/issues), @dblock could you please sync up one on this repo as well?

I made one in https://github.com/opensearch-project/dashboards-notifications/issues/31, but please don't hesitate to open issues for things like these next time and I can always link above.

**Comment by dblock** [comment]
> In this scheme, should maintainers should be able to add new maintainers to the repositories without intervention from the opensearch-project/admins? I thought yes, but I just reopened [opensearch-project/dashboards-search-relevance#146](https://github.com/opensearch-project/dashboards-search-relevance/issues/146) because I'm unable to access Settings -> Collaborators on the repo.

You're not. It's purposely centralized with the admin team to have some additional controls. I am working on a proposal to change that.

**Comment by dblock** [comment]
Verified all lists of MAINTAINERS, CODEOWNERS, and actual permissions. Removed all groups with more than triage access from all repos.
